File: FastStudy/models/StudySession.py
from typing import List, Optional
from datetime import datetime

from beanie import Document, Link
from pydantic import BaseModel, Field

class Task(BaseModel):
    name: str
    duration: int
    completed: bool = False

class StudySession(Document):
    # forward‐ref string, no import here
    user: Link["User"]

    tasks: List[Task]
    planned_duration: int
    actual_duration: Optional[int] = None

    start_time: datetime
    end_time: Optional[datetime] = None

    is_paused: bool = False
    paused_at: Optional[datetime] = None
    total_paused: int = Field(0, description="accumulated pause time, in seconds")
    last_heartbeat: datetime = Field(default_factory=datetime.utcnow)

    distractions_blocked: int = 0
    notes: Optional[str] = None

    class Settings:
        name = "study_sessions"

# at bottom, teach Pydantic about User
from models.User import User
StudySession.model_rebuild()
# to avoid circular import issues

File: vue-project/src/models/StudySession.ts
import type { Task } from './Task';

export interface RawStudySession {
  _id: string;
  user: string;                  // Link<User> comes back as an OID string
  tasks: Task[];
  planned_duration: number;
  actual_duration?: number;
  start_time: string;
  end_time?: string;
  is_paused: boolean;
  paused_at?: string;
  total_paused: number;          // seconds
  last_heartbeat: string;
  distractions_blocked: number;
  notes?: string;
}

export interface StudySession {
  _id: string;
  userId: string;
  tasks: Task[];
  plannedDuration: number;
  actualDuration?: number;
  startTime: string;
  endTime?: string;
  isPaused: boolean;
  pausedAt?: string;
  totalPaused: number;
  lastHeartbeat: string;
  distractionsBlocked: number;
  notes?: string;
}